src/NHibernate/Loader/Loader.cs

```diff
@@ -972,6 +972,35 @@ private void CheckVersion(int i, IEntityPersister persister, object id, object e
 					entry.LockMode = lockMode;
 				}
 			}
+
+			//save object to EntityUnique cache
+			//see LoadFromResultSet func
+			IAssociationType[] ownerAssociationTypes = OwnerAssociationTypes;
+			if (ownerAssociationTypes != null && ownerAssociationTypes[i] != null)
+			{
+				string ukName = ownerAssociationTypes[i].RHSUniqueKeyPropertyName;
+				if (ukName != null)
+				{
+					int index = ((IUniqueKeyLoadable) persister).GetPropertyIndex(ukName);
+					IType type = persister.PropertyTypes[index];
+
+					object id = key.Identifier;
+					bool eagerPropertyFetch = IsEagerPropertyFetchEnabled(i);
+					ILoadable entityPersister = (ILoadable) persister;
+					string[][] cols = EntityAliases[i].SuffixedPropertyAliases;
+
+					object[] values = entityPersister.Hydrate(rs, id, obj, entityPersister, cols, eagerPropertyFetch, session);
```

Comment:
Here, there is a trouble. The referenced property may be mutable. If it has changed in DB but is still loaded in session with the old value, we are then going to associate a unique key on the new value with an entity still holding the old value. Instead of reading its value from the result set, it should be read from the entity in session (`obj`).

Reply:
We need to test this case.

Reply:
Done in #1492.